onlinejudge/service/atcoder.py

```diff
@@ -636,29 +636,39 @@ def _get_tag_lang(cls, tag: bs4.Tag) -> Optional[str]:
 
     @classmethod
     def _find_sample_tags(cls, soup: bs4.BeautifulSoup) -> Iterator[Tuple[bs4.Tag, bs4.Tag]]:
-        for pre in soup.find_all('pre'):
-            log.debug('pre tag: %s', str(pre))
-            if not pre.string:
-                if len(pre.contents):
-                    # <pre> element is not a sample but the input format
-                    continue
-                # the sample is an empty string, so set pre.string as '' instead of None
-                pre.string = ''
-
-            def h3_plus(tag):
-                prv = tag.find_previous_sibling()
-                if prv and prv.name == 'h3' and prv.string:
-                    yield (pre, prv)
-
-            # the first format: h3+pre
-            yield from h3_plus(pre)
-
+        def tag_plus(tag, expected_prv, expected_strings):
+            prv = tag.find_previous_sibling()
+            if prv and prv.name == expected_prv and prv.string and any(s in prv.string for s in expected_strings):
+                yield (pre, prv)
+
+        expected_strings = ['入力例', '出力例', 'Sample Input', 'Sample Output']
+        finished = False
+
+        if not finished:
+            # the first format: p+pre
+            # this format uses 'literal-block' in its page
+            # example: https://atcoder.jp/contests/utpc2011/tasks/utpc2011_1
+            for pre in soup.find_all('pre', 'literal-block'):
+                finished = True
+                log.debug('pre tag: %s', str(pre))
+                yield from tag_plus(tag=pre, expected_prv='p', expected_strings=expected_strings)
+
+        if not finished:
             # the second format: h3+section pre
-            if pre.parent and pre.parent.name == 'section':
-                # ignore tags which are not samples
-                # example: https://atcoder.jp/contests/abc003/tasks/abc003_4
-                if pre.find_previous_sibling('pre') is None:
-                    yield from h3_plus(pre.parent)
+            # this format uses 'prettyprint linenums' in its page
+            # example: https://atcoder.jp/contests/abc003/tasks/abc003_4
+            for pre in soup.find_all('pre', 'prettyprint linenums'):
+                finished = True
+                log.debug('pre tag: %s', str(pre))
+                yield from tag_plus(tag=pre.parent, expected_prv='h3', expected_strings=expected_strings)
+
+        if not finished:
+            # the third format: h3+pre
+            # example: https://atcoder.jp/contests/abc114/tasks/abc114_d
+            for pre in soup.find_all('pre'):
+                finished = True
+                log.debug('pre tag: %s', str(pre))
+                yield from tag_plus(tag=pre, expected_prv='h3', expected_strings=expected_strings)
```

Comment:
`raise StopIteration` can remove `finished` flag and reduce nest depth.

Reply:
If we sort these branch by major branch first, this part will become more readable.
I mean
1. https://atcoder.jp/contests/abc114/tasks/abc114_d (current ABC)
2. https://atcoder.jp/contests/abc003/tasks/abc003_4 (old ABC)
3. https://atcoder.jp/contests/utpc2011/tasks/utpc2011_1 (utpc2011 only,)

Reply:
> If we sort these branch by frequency descending order (a.k.a major branch comes first), this part will become more readable.

I have implemented this as (old -> current) because it is easier for me to implement the logic than (current -> old) flow, but I agree with your opinion. I re-implement the collecting logic.  

<br/>

> Use raise StopIteration can remove finished flag and reduce nest level. How do you think? 

I'll try it.

Reply:
Sorry, I misunderstood your code.
In this situation, origin order is correct to maintain code simple.
And StopIteration does not meaningful.

How about
```python
        if soup.find('pre', 'literal-block'):
            # the first format: p+pre
            # this format uses 'literal-block' in its page
            # example: https://atcoder.jp/contests/utpc2011/tasks/utpc2011_1
            for pre in soup.find_all('pre', 'literal-block'):
                log.debug('pre tag: %s', str(pre))
                yield from tag_plus(tag=pre, expected_prv='p', expected_strings=expected_strings)
        elif soup.find('pre', 'prettyprint linenums'):
            # the second format: h3+section pre
            # this format uses 'prettyprint linenums' in its page
            # example: https://atcoder.jp/contests/abc003/tasks/abc003_4
            for pre in soup.find_all('pre', 'prettyprint linenums'):
                log.debug('pre tag: %s', str(pre))
                yield from tag_plus(tag=pre.parent, expected_prv='h3', expected_strings=expected_strings)
        elif soup.find('pre'):
            # the third format: h3+pre
            # example: https://atcoder.jp/contests/abc114/tasks/abc114_d
            for pre in soup.find_all('pre'):
                log.debug('pre tag: %s', str(pre))
                yield from tag_plus(tag=pre, expected_prv='h3', expected_strings=expected_strings)
```
?

I want not to use control flag like `finish`.
In many case, it make code complicated.

Reply:
What a Beautiful Code...!
I follow your advice!